File: backend/crud/player_state.py
"""
CRUD operations for PlayerState entities.

Handles player stats, inventory, action history, and current location tracking.
Business rules (clamping, merging) are delegated to domain/services/player_rules.py.
"""

import json
import logging
from typing import List, Optional

import models
import schemas
from database import serialized_write
from domain.services.player_rules import (
    InventoryItem as DomainInventoryItem,
)
from domain.services.player_rules import (
    apply_stat_changes,
    initialize_stats_from_definitions,
    merge_inventory_item,
)
from domain.services.player_rules import (
    remove_inventory_item as remove_item_from_list,
)
from domain.services.player_state_serializer import PlayerStateSerializer
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload

logger = logging.getLogger("PlayerStateCRUD")


async def get_player_state(
    db: AsyncSession,
    world_id: int,
) -> Optional[models.PlayerState]:
    """Get the player state for a world."""
    result = await db.execute(
        select(models.PlayerState)
        .options(selectinload(models.PlayerState.current_location))
        .where(models.PlayerState.world_id == world_id)
    )
    return result.scalar_one_or_none()


async def set_current_location(
    db: AsyncSession,
    world_id: int,
    location_id: int,
) -> Optional[models.PlayerState]:
    """Set the player's current location."""
    # Get player state
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return None

    # Clear is_current on old location
    if player_state.current_location_id:
        old_loc_result = await db.execute(
            select(models.Location).where(models.Location.id == player_state.current_location_id)
        )
        old_location = old_loc_result.scalar_one_or_none()
        if old_location:
            old_location.is_current = False

    # Set is_current on new location
    new_loc_result = await db.execute(select(models.Location).where(models.Location.id == location_id))
    new_location = new_loc_result.scalar_one_or_none()
    if new_location:
        new_location.is_current = True
        new_location.is_discovered = True  # Visiting discovers a location

    # Update player state
    player_state.current_location_id = location_id

    async with serialized_write():
        await db.commit()

    await db.refresh(player_state, attribute_names=["current_location"])
    return player_state


async def increment_turn(
    db: AsyncSession,
    world_id: int,
) -> int:
    """Increment the turn counter and return new value."""
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return 0

    player_state.turn_count += 1

    async with serialized_write():
        await db.commit()

    return player_state.turn_count


async def update_stats(
    db: AsyncSession,
    world_id: int,
    changes: dict,
    stat_definitions: Optional[dict] = None,
) -> dict:
    """Update player stats and return new values."""
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return {}

    # Parse current stats and apply changes with clamping (delegated to domain rules)
    current_stats = PlayerStateSerializer.parse_stats(player_state.stats)
    new_stats = apply_stat_changes(current_stats, changes, stat_definitions)

    # Save
    player_state.stats = PlayerStateSerializer.serialize_stats(new_stats)

    async with serialized_write():
        await db.commit()

    return new_stats


async def add_inventory_item(
    db: AsyncSession,
    world_id: int,
    item: schemas.InventoryItem,
) -> List[dict]:
    """Add an item to inventory."""
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return []

    # Parse current inventory and merge item (delegated to domain rules)
    inventory = PlayerStateSerializer.parse_inventory(player_state.inventory)
    domain_item = DomainInventoryItem(
        id=item.id,
        name=item.name,
        description=item.description,
        quantity=item.quantity,
        properties=item.properties,
    )
    new_inventory = merge_inventory_item(inventory, domain_item)

    # Save
    player_state.inventory = PlayerStateSerializer.serialize_inventory(new_inventory)

    async with serialized_write():
        await db.commit()

    return new_inventory


async def remove_inventory_item(
    db: AsyncSession,
    world_id: int,
    item_id: str,
    quantity: int = 1,
) -> tuple[bool, int]:
    """Remove an item from inventory. Returns (success, remaining_quantity)."""
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return False, 0

    # Parse current inventory and remove item (delegated to domain rules)
    inventory = PlayerStateSerializer.parse_inventory(player_state.inventory)
    new_inventory, success, remaining = remove_item_from_list(inventory, item_id, quantity)

    if not success:
        return False, remaining

    # Save
    player_state.inventory = PlayerStateSerializer.serialize_inventory(new_inventory)

    async with serialized_write():
        await db.commit()

    return True, remaining


async def add_action_to_history(
    db: AsyncSession,
    world_id: int,
    turn: int,
    action: str,
    result_text: str,
) -> None:
    """Add an action to the player's action history."""
    db_result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = db_result.scalar_one_or_none()

    if not player_state:
        return

    # Parse current history
    history = json.loads(player_state.action_history) if player_state.action_history else []

    # Add new action
    history.append(
        {
            "turn": turn,
            "action": action,
            "result": result_text,
        }
    )

    # Keep only last 10 actions
    history = history[-10:]

    # Save
    player_state.action_history = PlayerStateSerializer.serialize_action_history(history)

    async with serialized_write():
        await db.commit()


async def initialize_player_stats(
    db: AsyncSession,
    world_id: int,
    stat_definitions: dict,
    initial_stats: Optional[dict] = None,
) -> dict:
    """Initialize player stats based on definitions."""
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        return {}

    # Build default stats from definitions (delegated to domain rules)
    new_stats = initialize_stats_from_definitions(stat_definitions, initial_stats)

    # Save
    player_state.stats = PlayerStateSerializer.serialize_stats(new_stats)

    async with serialized_write():
        await db.commit()

    return new_stats


# =============================================================================
# Chat Mode Operations
# =============================================================================


async def enter_chat_mode(
    db: AsyncSession,
    world_id: int,
    start_message_id: int,
) -> Optional[int]:
    """
    Enter chat mode for a world.

    Args:
        db: Database session
        world_id: World ID
        start_message_id: Message ID marking the start of chat mode

    Returns:
        The new chat_session_id if successfully entered chat mode, None if already in chat mode or error
    """
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        logger.warning(f"Cannot enter chat mode: PlayerState not found for world {world_id}")
        return None

    if player_state.is_chat_mode:
        logger.info(f"Already in chat mode for world {world_id}")
        return None

    # Generate a new chat session ID using timestamp-based approach
    import time

    chat_session_id = int(time.time() * 1000) % (2**31 - 1)  # Milliseconds, capped for int

    player_state.is_chat_mode = True
    player_state.chat_mode_start_message_id = start_message_id
    player_state.chat_session_id = chat_session_id

    async with serialized_write():
        await db.commit()

    logger.info(
        f"Entered chat mode for world {world_id}, start_message_id={start_message_id}, chat_session_id={chat_session_id}"
    )
    return chat_session_id


async def exit_chat_mode(
    db: AsyncSession,
    world_id: int,
) -> Optional[tuple[int, int]]:
    """
    Exit chat mode for a world.

    Args:
        db: Database session
        world_id: World ID

    Returns:
        Tuple of (start_message_id, chat_session_id) if was in chat mode, None if not in chat mode or error
    """
    result = await db.execute(select(models.PlayerState).where(models.PlayerState.world_id == world_id))
    player_state = result.scalar_one_or_none()

    if not player_state:
        logger.warning(f"Cannot exit chat mode: PlayerState not found for world {world_id}")
        return None

    if not player_state.is_chat_mode:
        logger.info(f"Not in chat mode for world {world_id}")
        return None

    start_message_id = player_state.chat_mode_start_message_id
    chat_session_id = player_state.chat_session_id
    player_state.is_chat_mode = False
    # Keep chat_mode_start_message_id so frontend can use it as resume point
    # It will be overwritten when entering chat mode again
    player_state.chat_session_id = None

    async with serialized_write():
        await db.commit()

    logger.info(
        f"Exited chat mode for world {world_id}, start_message_id={start_message_id}, chat_session_id={chat_session_id}"
    )
    return (start_message_id, chat_session_id)

File: backend/src/crud/player-state.ts
/**
 * CRUD for PlayerState: read the row, hand the JSON blob to
 * `domain/player-rules.ts`, write it back. Those rules are shared with
 * `services/player-service.ts`, which runs the same mutations against
 * `player.json`, so none of the logic is inlined here.
 */

import { eq } from 'drizzle-orm'
import type { Db } from '../db'
import { locations, playerStates, type Location, type PlayerState } from '../db/schema'
import {
  applyStatChanges,
  initializeStatsFromDefinitions,
  InventoryItem,
  mergeInventoryItem,
  removeInventoryItem as removeItemFromList,
  type InventoryEntry,
  type InventoryItemFields,
  type StatDefinitions,
} from '../domain/player-rules'
import { PlayerStateSerializer } from '../domain/player-state-serializer'
import { getLogger } from '../infrastructure/logging/logger'

const logger = getLogger('PlayerStateCRUD')

export interface PlayerStateWithLocation extends PlayerState {
  currentLocation: Location | null
}

/** How many entries `action_history` keeps. Older ones are dropped on write. */
const ACTION_HISTORY_LIMIT = 10

export interface ActionHistoryEntry {
  turn: number
  action: string
  result: string
}

// `world_id` is UNIQUE, so this is always at most one row.
function loadState(db: Db, worldId: number): PlayerState | null {
  return db.select().from(playerStates).where(eq(playerStates.worldId, worldId)).get() ?? null
}

export function getPlayerState(db: Db, worldId: number): PlayerStateWithLocation | null {
  const row = db
    .select({ state: playerStates, currentLocation: locations })
    .from(playerStates)
    .leftJoin(locations, eq(playerStates.currentLocationId, locations.id))
    .where(eq(playerStates.worldId, worldId))
    .get()

  return row ? { ...row.state, currentLocation: row.currentLocation } : null
}

/**
 * Move the player to a location. Three writes in one transaction: a poll landing
 * between the first and third would render a map with two current locations or
 * none. Visiting also *discovers* the location, which is what makes somewhere
 * the `travel` tool moved to appear in the player's list.
 */
export function setCurrentLocation(
  db: Db,
  worldId: number,
  locationId: number,
): PlayerStateWithLocation | null {
  const state = loadState(db, worldId)
  if (!state) return null

  db.transaction(() => {
    if (state.currentLocationId !== null) {
      db.update(locations)
        .set({ isCurrent: false })
        .where(eq(locations.id, state.currentLocationId))
        .run()
    }

    db.update(locations)
      .set({ isCurrent: true, isDiscovered: true })
      .where(eq(locations.id, locationId))
      .run()

    db.update(playerStates)
      .set({ currentLocationId: locationId })
      .where(eq(playerStates.worldId, worldId))
      .run()
  })

  return getPlayerState(db, worldId)
}

// 0 when the world has no player state; callers treat the turn number as
// advisory display data. `turn_count` is nullable, hence the fold.
export function incrementTurn(db: Db, worldId: number): number {
  const state = loadState(db, worldId)
  if (!state) return 0

  const next = (state.turnCount ?? 0) + 1
  db.update(playerStates).set({ turnCount: next }).where(eq(playerStates.worldId, worldId)).run()
  return next
}

// The cast is deliberate: unlike `services/player-service.ts`, nothing is
// filtered out — that would silently delete stats from the stored column.
function readStats(state: PlayerState): Record<string, number> {
  return PlayerStateSerializer.parseStats(state.stats) as Record<string, number>
}

// Omitting `statDefinitions` means unbounded stats — how a stat an agent
// invented mid-game behaves.
export function updateStats(
  db: Db,
  worldId: number,
  changes: Record<string, number>,
  statDefinitions?: StatDefinitions | null,
): Record<string, number> {
  const state = loadState(db, worldId)
  if (!state) return {}

  const newStats = applyStatChanges(readStats(state), changes, statDefinitions)

  db.update(playerStates)
    .set({ stats: PlayerStateSerializer.serializeStats(newStats) })
    .where(eq(playerStates.worldId, worldId))
    .run()

  return newStats
}

/**
 * Build the starting stat block from the world's definitions. This *replaces*
 * the blob rather than merging, so a second call mid-campaign wipes the
 * player's progress; character creation is the only caller.
 */
export function initializePlayerStats(
  db: Db,
  worldId: number,
  statDefinitions: StatDefinitions,
  initialStats?: Record<string, number> | null,
): Record<string, number> {
  const state = loadState(db, worldId)
  if (!state) return {}

  const newStats = initializeStatsFromDefinitions(statDefinitions, initialStats)

  db.update(playerStates)
    .set({ stats: PlayerStateSerializer.serializeStats(newStats) })
    .where(eq(playerStates.worldId, worldId))
    .run()

  return newStats
}

/**
 * Add an item to the inventory. The stored shape is the embedded one, *not* the
 * reference shape `player.json` uses — the column holds the full item.
 */
export function addInventoryItem(
  db: Db,
  worldId: number,
  item: InventoryItemFields,
): InventoryEntry[] {
  const state = loadState(db, worldId)
  if (!state) return []

  const inventory = PlayerStateSerializer.parseInventory<InventoryEntry>(state.inventory)
  const newInventory = mergeInventoryItem(inventory, new InventoryItem(item))

  db.update(playerStates)
    .set({ inventory: PlayerStateSerializer.serializeInventory(newInventory) })
    .where(eq(playerStates.worldId, worldId))
    .run()

  return newInventory
}

export interface RemoveInventoryItemResult {
  success: boolean
  /** Quantity left afterwards; on a shortfall, the quantity actually held. */
  remaining: number
}

/**
 * Remove `quantity` of an item. Nothing is written when the removal fails,
 * shortfalls included, so a caller charging an item cost cannot take partial
 * payment; `remaining` is then what the player actually holds.
 */
export function removeInventoryItem(
  db: Db,
  worldId: number,
  itemId: string,
  quantity = 1,
): RemoveInventoryItemResult {
  const state = loadState(db, worldId)
  if (!state) return { success: false, remaining: 0 }

  const inventory = PlayerStateSerializer.parseInventory<InventoryEntry>(state.inventory)
  const result = removeItemFromList(inventory, itemId, quantity)

  if (!result.success) return { success: false, remaining: result.remaining }

  db.update(playerStates)
    .set({ inventory: PlayerStateSerializer.serializeInventory(result.inventory) })
    .where(eq(playerStates.worldId, worldId))
    .run()

  return { success: true, remaining: result.remaining }
}

// A corrupt blob throws rather than silently resetting the history; quietly
// discarding a player's recent actions is the worse failure.
export function addActionToHistory(db: Db, worldId: number, entry: ActionHistoryEntry): void {
  const state = db
    .select({ actionHistory: playerStates.actionHistory })
    .from(playerStates)
    .where(eq(playerStates.worldId, worldId))
    .get()

  if (!state) return

  const history: unknown = state.actionHistory ? JSON.parse(state.actionHistory) : []
  if (!Array.isArray(history)) {
    throw new TypeError(`player_states.action_history for world ${worldId} is not a JSON array`)
  }

  history.push({ turn: entry.turn, action: entry.action, result: entry.result })

  db.update(playerStates)
    .set({ actionHistory: JSON.stringify(history.slice(-ACTION_HISTORY_LIMIT)) })
    .where(eq(playerStates.worldId, worldId))
    .run()
}

// Chat mode and gameplay share a room and the `messages` table, and
// `chat_session_id` is the whole mechanism keeping the two transcripts apart.
// The next two functions own that id's lifetime.

/**
 * Enter chat mode, minting the session id every chat message will carry. `null`
 * both when there is no player state and when already in chat mode: either way
 * the caller must not start a second session, and minting a fresh id would
 * strand the running session's messages under the old one.
 */
export function enterChatMode(db: Db, worldId: number, startMessageId: number): number | null {
  const state = loadState(db, worldId)
  if (!state) {
    logger.warning(`Cannot enter chat mode: PlayerState not found for world ${worldId}`)
    return null
  }

  if (state.isChatMode) {
    logger.info(`Already in chat mode for world ${worldId}`)
    return null
  }

  const chatSessionId = Date.now() % (2 ** 31 - 1)

  db.update(playerStates)
    .set({
      isChatMode: true,
      chatModeStartMessageId: startMessageId,
      chatSessionId,
    })
    .where(eq(playerStates.worldId, worldId))
    .run()

  logger.info(
    `Entered chat mode for world ${worldId}, start_message_id=${startMessageId}, chat_session_id=${chatSessionId}`,
  )
  return chatSessionId
}

export interface ExitChatModeResult {
  /** The message the session started from; the frontend resumes rendering here. */
  startMessageId: number | null
  chatSessionId: number | null
}

/**
 * Leave chat mode, reporting the session that just ended.
 * `chat_mode_start_message_id` is deliberately *not* cleared — the frontend
 * reads it after the exit as where to resume the gameplay transcript.
 * `chat_session_id` *is*, or the next message would look like part of it.
 */
export function exitChatMode(db: Db, worldId: number): ExitChatModeResult | null {
  const state = loadState(db, worldId)
  if (!state) {
    logger.warning(`Cannot exit chat mode: PlayerState not found for world ${worldId}`)
    return null
  }

  if (!state.isChatMode) {
    logger.info(`Not in chat mode for world ${worldId}`)
    return null
  }

  const startMessageId = state.chatModeStartMessageId
  const chatSessionId = state.chatSessionId

  db.update(playerStates)
    .set({ isChatMode: false, chatSessionId: null })
    .where(eq(playerStates.worldId, worldId))
    .run()

  logger.info(
    `Exited chat mode for world ${worldId}, start_message_id=${startMessageId}, chat_session_id=${chatSessionId}`,
  )
  return { startMessageId, chatSessionId }
}
